pkg/services/scm_provider/github.go

```diff
@@ -63,22 +84,14 @@ func (g *GithubProvider) ListRepos(ctx context.Context, cloneProtocol string) ([
 			default:
 				return nil, fmt.Errorf("unknown clone protocol for GitHub %v", cloneProtocol)
 			}
-
-			branches, err := g.listBranches(ctx, githubRepo)
-			if err != nil {
-				return nil, fmt.Errorf("error listing branches for %s/%s: %v", githubRepo.Owner.GetLogin(), githubRepo.GetName(), err)
-			}
-
-			for _, branch := range branches {
-				repos = append(repos, &Repository{
-					Organization: githubRepo.Owner.GetLogin(),
-					Repository:   githubRepo.GetName(),
-					URL:          url,
-					Branch:       branch.GetName(),
-					SHA:          branch.GetCommit().GetSHA(),
-					Labels:       githubRepo.Topics,
-				})
-			}
+			repos = append(repos, &Repository{
+				Organization: githubRepo.Owner.GetLogin(),
+				Repository:   githubRepo.GetName(),
+				Branch:       githubRepo.GetDefaultBranch(),
```

Comment:
`GetDefaultBranch()` would still call Github API for every repo even if there is a `repositoryMatch` or `labelMatch` filter.

At this point, we already do know the repo names and their labels/topics (from `ListByOrg`). So why not match the names and labels/topics with `repositoryMatch` or `labelMatch` and skip calling `GetDefaultBranch()` alltogether?

Reply:
As for the other filters (`pathsExist` and `branchMatch`), there is no way around it. Github API must be called for every repo to get that data 😞

Reply:
ok, thanks for that insight, I'll update that to use the default branch that is already present, as opposed to getting the default branch with another API call.  I didn't realize it was doing that.

Reply:
In regards to the two filters you mention, PathExists, and BranchMatch, those two filters will get applied now at two different phases.   The new function in utils.go will determine where a filter is for a branch, or for a repo.  PathExists is meant for a repo, and gets applied during ListRepos function.  The other filter, BranchMatch, gets applied only after we've applied other filters to the repos, whittling down the repos using either PathExists, or LabelMatch.  

Here is that function, which determines if a filter is for a branch, or repo:

```
func getApplicableFilters(filters []*Filter, branchFilter bool) []*Filter {
	applicableFilters := []*Filter{}
	for _, filter := range filters {
		if branchFilter && filter.BranchMatch != nil {
			applicableFilters = append(applicableFilters, filter)
		} else if !branchFilter && filter.BranchMatch == nil {
			applicableFilters = append(applicableFilters, filter)
		}
	}
	return applicableFilters
}
```

Reply:
Thank you very much for the explanation, it's clear now. Solid work 👍🏼

Reply:
I think applying `PathsExist` at the repo-level filter might be a mistake. Even if the default branch doesn't have a certain path, a non-default branch might have that path.

Reply:
I moved this to a branch level filter